### DISCUSSION TITLE
How to host your bookmarks on Github? error 401

### DISCUSSION THREAD

**Discussion by oeloo** [floccusaddon/floccus#1655] [Q&A] [answered]
Hi, I have created a repository MyBookmarks on Github:
![image](https://github.com/floccusaddon/floccus/assets/98819304/c407f3dc-e6d5-4097-8aa5-c0e28090174a)
and entered the URL and login/password:
![image](https://github.com/floccusaddon/floccus/assets/98819304/3dbdd490-dba3-47d3-8dce-3548abe0a857)
but it is not working:
![image](https://github.com/floccusaddon/floccus/assets/98819304/97fe2f6e-b699-4976-b3a7-19b80c7f8c70)
As far as I know GitHub requires 2FA, so how to host your bookmarks on Github? regards

**Comment by marcelklehr** [comment]
Hey @oeloo

Great question. You'll need a personal access token to cricumvent 2FA. Go to GitHub -> Account Settings -> Developer Settings -> Personal Access tokens and you can create one.

**Reply by oeloo** [reply]
Thank you, that makes sense. When creating a Personal Access Token for Floccus, what is the minimum scope needed by Floccus (to limit security exposure)?

**Reply by marcelklehr** [reply]
It only needs push (write) access to the one repository you want to sync with

**Reply by oeloo** [reply]
I have created a token (classic) and used it in the Floccus configuration:
login=<my email (also tested with my GitHub's username)>
password=<the generated token (ghp_xxxxxxxxxxxxxxxxxxxxxxxxxxx) >
![image](https://github.com/floccusaddon/floccus/assets/98819304/1f2d4c2d-317b-4a13-8f70-4c1b954fd03b)

But it does not work, even when I give all the scope access: I get "HTTP Error: 403 Last synchronized":
![image](https://github.com/floccusaddon/floccus/assets/98819304/a5735456-9484-494f-b98f-cd3776b483ff)

> It only needs push access to the one repository you want to sync with

I do not see the push access in the scope list:
![image](https://github.com/floccusaddon/floccus/assets/98819304/394c294b-cc8a-42d9-930d-8c3a98663de2)

**Reply by marcelklehr** [reply]
In the classic access tokens you cannot select individual repositories, so you will have to select full access to private repos. ~~You might need to add `.git` to the end of the URL since that is the git HTTP URL~~

**Reply by oeloo** [reply]
I have created a 2nd token, a fine-grained one: github_pat_xxxxxxxxxxxxx_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
but when I enter it in the Floccus configuration, I still have an error:
![image](https://github.com/floccusaddon/floccus/assets/98819304/ba88659c-afe0-478d-9f1e-004a19db4399)
![image](https://github.com/floccusaddon/floccus/assets/98819304/85a6ab8b-df96-4448-9f5b-e852b5f43c4f)

Why is not working? I have given ALL the permissions on my repository, as you can see:
![image](https://github.com/floccusaddon/floccus/assets/98819304/a946477c-0bd3-49eb-9939-7487ea7a0d07)

> It only needs push access to the one repository you want to sync with

I do not see a push access permission in the fine-grained token as well.

**Reply by marcelklehr** [reply]
> Could not find origin/main

This is a different error though. The branch you want to use must already exist. When creating a repo on github, the default branch will be master, if you want to sync to main, you need to manually create a main branch.

**Comment by oeloo** [comment]
You're right, the branch was the issue. Sync is now working between floccus addon for Chrome on Windows 😊:
all my Chrome bookmarks:
![image](https://github.com/floccusaddon/floccus/assets/98819304/25bee668-3af8-46b1-9ef7-ad4c96751f4d)
are now sync to Github (I have checked the .xbel its content is OK):
![image](https://github.com/floccusaddon/floccus/assets/98819304/3784ec17-0aa5-4797-acc9-772451274264)

But when I make the exact same configuration (I have even tested several other times with importing config from floccus on Chrome) on floccus addon for Kiwi on Android (which initially had no bookmarks), the sync works but only a part of the bookmarks are present: I only have the bookmarks inside my folder 'Mobile bookmarks', all the bookmarks inside the folder 'Other bookmarks' are missing! (these bookmarks are present in the .xbel on Github, I have fully checked)
![image](https://github.com/floccusaddon/floccus/assets/98819304/a0177b1b-86e6-4ff0-a688-14f756ae37aa)

And when I make the exact same configuration (I have even tested several other times with importing config from floccus on Chrome) on floccus Android app: the sync does not run and gives "HTTP Error: 400"
![image](https://github.com/floccusaddon/floccus/assets/98819304/c1bec1cf-c833-4c78-8ed9-b0117fca2890)

How it is possible that floccus behaves so differently with the exact same settings?

**Reply by marcelklehr** [reply, marked as answer]
> I only have the bookmarks inside my folder 'Mobile bookmarks', all the bookmarks inside the folder 'Other bookmarks' are missing! 

You'll need to select the Mobile Bookmarks folder as the folder to sync instead of the root folder. Most browsers do not allow creating top-level folders.

The git sync is currently broken on the native app, sadly, see #1608

**Reply by oeloo** [reply]
> You'll need to select the Mobile Bookmarks folder as the folder to sync instead of the root folder. Most browsers do not allow creating top-level folders.

I confirm this setting works, all bookmark folders are on Kiwi now.
What is the minimum permission settings I should give to Personal Access Token? there is no mention of push in the settings proposed by GitHub.

> The git sync is currently broken on the native app, sadly, see #1608

I understand now. While this issue is not solved on the framework you need to use, I suggest you remove Git from the list of sync possibilities inside floccus app on Android (as it is misleading for users willing to use Git as sync).

**Reply by oeloo** [reply]
What minimum permission should we set in GitHub for floccus to run?
![image](https://github.com/floccusaddon/floccus/assets/98819304/595dd2d3-98bb-4cb8-88b1-063b22cd8079)
I have tried removing the user permissions, but this is not working as floccus is then not syncing.

**Reply by marcelklehr** [reply]
If you select a specific repository to give access to, the following permissions should be enough:

![image](https://github.com/floccusaddon/floccus/assets/986878/3a3272b5-6c72-4144-81fa-c88fb9abfcd9)

**Reply by oeloo** [reply]
Thank you for the permissions, I have it is enough and it works :)

**Comment by balthanon** [comment]
This was a really useful issue to refer back to when this 401 error occurred for me.  I did want to add a quick note to it that the Personal Access Token may be needed even when two factor authentication isn't set up-- I don't have it in place on my account since I don't have much stored in it (though I may change that now that I've set up my bookmarks here) and still ended up having to use the token.
